### PR TITLE
feat : composite types with new AST

### PR DESCRIPTION
**Pull request by mali-tintash**
- Tuples,
- tuples within tuples
- principal within tuples
- Lists
- optionals

Verified to be functional with new AST

Note : 
Addition of "attributes" field in the AST . Search of "parent_identifier" and you'll find comments on what it does.
But generally, the problem is that when we jump into get_expr, we are only working with the "value" node, and have no information of parent containing node. That usually contains the name of the variable.
This isn't required for most cases, but for tuples it was needed. Maybe we can review and it is no longer needed. But for now, it is implemented by passing parent information down using "attributes".

Things that need looking into : 
- There is some padding issue when we encapsulate a struct within a struct along some other members. This is specifically for when we pack tuple within tuples, or any other struct types inside a tuple struct. This needs to be looked into.
- Another thing i have noticed that for "tuple" specifically, the symbol that creates the type is showing the value assigned to it along with it's flag set to L-value, whereas they can only be "type".
- I am going to create two issues to represent these two items so they don't get ignored down the line.

Issues created :
https://github.com/proj-clarity-esbmc/clara_esbmc/issues/14
https://github.com/proj-clarity-esbmc/clara_esbmc/issues/13

### CODE REVIEW

File: src/clarity-frontend/clarity_convert.cpp [hunk 1]
```diff
@@ -4923,60 +4836,137 @@ bool clarity_convertert::get_array_to_pointer_type(
...
+    symbolt symbol;
+    get_default_symbol(symbol, debug_modulename, t, name, id, location_begin);
+    symbol.is_type = true;
+    added_symbol = move_symbol_to_context(symbol);
+    
+  }  
+  
+  
+  t = added_symbol->type;
+  t.tag("struct " + name);
```
Comment (mutahirtintash) [inline src/clarity-frontend/clarity_convert.cpp:4900]: Should tags has spaces. Some times string processing causes problems with spaces

File: src/clarity-frontend/clarity_convert.cpp [hunk 2]
```diff
@@ -4923,60 +4836,137 @@ bool clarity_convertert::get_array_to_pointer_type(
...
+        symbolt *nested_symbol = context.find_symbol(mem_id);
+        mem_type = nested_symbol->type;
+      }
+      else if (mem_type.get("#clar_type") == "principal")
+      {
+        //std::cout <<mem_type.pretty()<<std::endl;
+        mem_id = "tag-struct principal";
+        
+      }
+      else
```
Comment (mutahirtintash) [inline src/clarity-frontend/clarity_convert.cpp:4962]: Will this empty else be filled in later

File: src/clarity-frontend/clarity_convert.cpp (hunk 1 above)
Comment (mali-tintash) [reply src/clarity-frontend/clarity_convert.cpp:4900]: yes, tags have spaces. a tag looks like this "struct <struct name>" hence the string catenation is between 'struct ' (with a space at the end) + name .

File: src/clarity-frontend/clarity_convert.cpp (hunk 2 above)
Comment (mali-tintash) [reply src/clarity-frontend/clarity_convert.cpp:4962]: i'll fix it. it's supposed to be a No-op

Comment (mali-tintash) [reply src/clarity-frontend/clarity_convert.cpp:4962]: I'll fix later. this is a placeholder to add more types as we experiment more.

**Review by mutahirtintash — APPROVED** [review]

**Pull request merged**